pagopa/pn-frontend
# PN 1953 - FE show status and downtime log, allow to download downtime certification

## Short description
This PR includes all the needed stuff to show a page of app status and downtime log, including the possibility of downloading a legally valid certificate ("attestazione opponibile a terzi") for any downtime event.

## List of changes proposed in this pull request
- We define a model for the app status and for each downtime event.
- We implement the API calls for the status/downtime API provided by DXC, including validations performed through `pn-validator`. Is the first actual use of `pn-validator` in operative code. There are several tests regarding different possible cases of corrupted data coming from API. 
- We implemented the AppStatus page. We note the way in which we avoided duplication of logic between the mobile and desktop versions of the downtime log table.
- We enabled the use of the `pn-validator` package in both pf-webapp and pa-webapp, by adding the corresponding line in the `craco.config.js` files in each folder.
- We made some modifications to pn-common components
  - `ItemsCard` (used to show a list of elements in mobile mode), now the header can include either one or two elements. For the list of downtimes, we needed just one element in header, centered. To center the header we use the already available `headerGridProps` prop.
  - Also in `ItemsCard`, we fixed some bugs, specifically the calls to `getLabel` didn't pass the whole `Item` as parameter, so any definition of `getLabel` which actually needed the `Item` raised an error.
  - Still in `ItemsCard`, we added `data-testid="cardBodyValue"` for values not wrapped in typography, to make all values alike to this respect (previously, only the values wrapped in typography by `ItemsCard` carried this `data-testid`).
  - `EmptyState`: refactored the specification of which "sentiment" (sad / happy / custom) icon should be shown. The previous `disableSentimentDissatisfied` prop was replaced with a new `sentimentIcon`, which can be either a member of the `KnownSentiment` enum (options DISSATISFIED / SATISFIED / NONE) or a `SvgIconComponent`. In this way, we promote uniformity for the common cases, and at the same time, allow for arbitrary "sentiment icons". The default is `KnownSentiment.DISSATISFIED`, to preserve the default behavior (`disableSentimentDissatisfied = false`) of the previous version. <br/> I changed and manually tested all the uses of `EmptyState` in pa-webapp and pf-webapp.
  - Also in `EmptyState`, we adjusted the size of the "face" icon, either happy, sad or custom; this was agreed with the UI team (Norma and Alice).
  - `SideMenuListItem`: now the icon can be either a React component name (as up to now), or a function which yields the already instantiated icon. To achieve this, I changed the `SideMenuItem` interface. We needed this because the icon to show in the side menu for the app status option (as well as its color) depends on the actual status (OK or not-OK).
  - date services/utilities: we added two functions to format date/time, and an additional utility function. 
    - The new function `formatDateTime` yields an object `{ date, time }` with the strings that represent the corresponding parts of a timestamp; the time has the legend _"ore "_ prepended, e.g. _"ore 15:08"_. This legend _"ore"_ is i18n-ed.
    - In turn, `formatTimeHHMM`, yields the hour/minute part of a Date in HH:MM format, 
    - Finally, we also added a `minutesBeforeNow` function which yields a date in the past ... as the function name indicates.
  - Also regarding date services, the _"Oggi"_ legend issued by `formatDate` was i18n-ed.
  - Data regex: we added the `isoDate` regex.
  - We added a new `useDownloadDocument` hook that takes in charge the download of a document, given its URL. We used this hook in the new app status page. It could be used as well for the various downloads in the notification detail FE, I opened https://pagopa.atlassian.net/browse/PN-2708 to this respect.
- As indicated above, we actually implemented the first use of the `pn-validator` package. Some modifications of that package were needed. We also add the `IsOneOf` rule as part of this PR.

## How to test
Just launch the app (at the moment only pf, I should replicate the FE in pa next week), and look for a "Stato della piattaforma" option in the left-hand side menu. In DEV environment, this uses an actual API provided by DXC.  
If you want to variate the results, you can play with `AppStatus.api.ts`, there is a `useMockResponseData` boolean, just enable it for development (for test should be always false), the mock responses are in the same `AppStatus.api.ts` file.
